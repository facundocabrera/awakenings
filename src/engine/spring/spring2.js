// source https://www.math.arizona.edu/~jwatkins/h-ode.pdf

/**
 * Spring.
 * 
 * @param {Number} A Amplitude
 * @param {Number} k Spring constant
 * @param {Number} m Mass 
 */
function Spring(A, k, m) {
  const w = Math.sqrt( k / m );

  function y(t) {
    return A * Math.exp(-1 * t) * Math.cos(t);
  }

  function force(t) {
    return -1 * k * y(t);
  }

  return {
    y,
    force
  }
}

export {
  Spring
}
