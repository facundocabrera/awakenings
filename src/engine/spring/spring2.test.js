import { Spring } from "./spring2";

test('lets try our spring', () => {

  const amplitude = 10;
  const k = 16;     // spring k constant 
  const m = k / 2;  // a "mass"

  const σ = Spring(amplitude, k, m);

  const positions = [];
  const forces = [];

  let y = 1;
  let time = 0;

  // given this version is with decay, we iterate till we reach aprox 0
  // which means the spring is not moving anymore.
  while(y > 0.01) {
    y = σ.y(time);
    positions.push( y );
    forces.push( σ.force(time) );
    time += 0.1;
  }

  expect(positions).toMatchSnapshot();
  expect(forces).toMatchSnapshot();
});