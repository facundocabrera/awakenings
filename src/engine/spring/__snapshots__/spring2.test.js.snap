// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lets try our spring 1`] = `
Array [
  10,
  9.003169998451938,
  8.024106473425203,
  7.077306780263506,
  6.174056479016457,
  5.322807302156707,
  4.5295378914525,
  3.7980938992515383,
  3.130505040044801,
  2.527277532911687,
  1.9876611034641296,
  1.5098903250017768,
  1.0914005828987694,
  0.7290193539955684,
  0.41913381409028555,
  0.15783603136566277,
  -0.05895281879737542,
]
`;

exports[`lets try our spring 2`] = `
Array [
  -160,
  -144.050719975231,
  -128.38570357480324,
  -113.2369084842161,
  -98.78490366426331,
  -85.16491683450731,
  -72.47260626324,
  -60.76950238802461,
  -50.08808064071682,
  -40.43644052658699,
  -31.802577655426074,
  -24.15824520002843,
  -17.46240932638031,
  -11.664309663929094,
  -6.706141025444569,
  -2.5253765018506042,
  0.9432451007580067,
]
`;
